internal/wantrequestmanager/wantrequestmanager.go

```diff
@@ -0,0 +1,453 @@
+// WantRequestManager informs all Sessions that are interested in the message.
+//
+type WantRequestManager struct {
+	bstore blockstore.Blockstore
+	bpm    *bsbpm.BlockPresenceManager
+	lk     sync.RWMutex
+	wrs    map[cid.Cid]map[*WantRequest]struct{}
+}
+
+// New creates a new WantRequestManager
+func New(bstore blockstore.Blockstore, bpm *bsbpm.BlockPresenceManager) *WantRequestManager {
+	return &WantRequestManager{
+		bstore: bstore,
+		bpm:    bpm,
+		wrs:    make(map[cid.Cid]map[*WantRequest]struct{}),
+	}
+}
+
+// NewWantRequest creates a new WantRequest with the given set of wants.
+func (wrm *WantRequestManager) NewWantRequest(ks []cid.Cid) (*WantRequest, error) {
+	wr := newWantRequest(wrm, ks)
+
+	wrm.lk.Lock()
+
+	for _, c := range ks {
+		// Add the WantRequest to the set of WantRequests that are interested
+		// in this key
+		wrs, ok := wrm.wrs[c]
+		if !ok {
+			wrs = make(map[*WantRequest]struct{})
+			wrm.wrs[c] = wrs
+		}
+		wrs[wr] = struct{}{}
+	}
+
+	wrm.lk.Unlock()
+
+	// Check if any of the wanted keys are already in the blockstore (unlikely
+	// but possible) and if so, publish them to all WantRequests that want them
+	blks, err := wrm.getBlocks(ks)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(blks) > 0 {
+		wrm.publish(&IncomingMessage{Blks: blks})
+	}
+
+	return wr, nil
+}
+
+// getBlocks gets the blocks corresponding to the given keys from the
+// blockstore
+func (wrm *WantRequestManager) getBlocks(ks []cid.Cid) ([]blocks.Block, error) {
+	var blks []blocks.Block
+	for _, c := range ks {
+		// TODO: do this in parallel?
+		b, err := wrm.bstore.Get(c)
+		if err == blockstore.ErrNotFound {
+			continue
+		}
+
+		if err != nil {
+			return nil, err
+		}
+
+		blks = append(blks, b)
+	}
+	return blks, nil
+}
+
+// PublishToSessions sends the message to all sessions that are interested in
+// it
+func (wrm *WantRequestManager) PublishToSessions(msg *IncomingMessage) (*cid.Set, error) {
+	local := msg.From == ""
+
+	// If the message includes blocks
+	if len(msg.Blks) > 0 {
+		wanted := msg.Blks
+		if !local {
+			// If the blocks came from the network, only put blocks to the
+			// blockstore if the local node actually wanted them
+			wanted = wrm.wantedBlocks(msg.Blks)
+		}
+
+		if len(wanted) > 0 {
+			// Put blocks to the blockstore
+			err := wrm.bstore.PutMany(wanted)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// Publish the message to WantRequests that are interested in the
+	// blocks / HAVEs / DONT_HAVEs in the message
+	return wrm.publish(msg), nil
+}
+
+// Returns blocks that are wanted by one of the sessions
+func (wrm *WantRequestManager) wantedBlocks(blks []blocks.Block) []blocks.Block {
+	wrm.lk.Lock()
+	defer wrm.lk.Unlock()
+
+	wanted := make([]blocks.Block, 0, len(blks))
+	for _, b := range blks {
+		c := b.Cid()
+
+		// Check if the block is wanted by one of the sessions
+		for wr := range wrm.wrs[c] {
+			if wr.wants(c) {
```

Comment:
We could have 2 maps but that would mean when a want changes state from wanted -> unwanted we have to update both maps, so I believe it would be less performant

Reply:
Updating the wants might be slightly more expensive, but we'd be able to get rid of this loop. An extra map operation per block received is better than iterating over the entire wantlist.

Reply:
Ah so this code is actually iterating over all the WantRequests who have registered interest in the want (there will almost always be just one).

To clarify:
```
func (wrm *WantRequestManager) wantedBlocks(blks []blocks.Block) []blocks.Block {
	wrm.lk.Lock()
	defer wrm.lk.Unlock()

	wanted := make([]blocks.Block, 0, len(blks))
	for _, b := range blks {
		c := b.Cid()

		// Check if the block is wanted by one of the sessions
		for wr := range wrm.wrs[c] {
			if wr.wants(c) {
				wanted = append(wanted, b)
				break
			}
		}
	}
	return wanted
}

```

`wrm.wrs` is a mapping from `CID -> <set of WantRequests>`, so
```
for wr := range wrm.wrs[c] {
```
iterates over `<set of WantRequests>`